### PR TITLE
Add a decompileCFR task

### PR DESCRIPTION
**Pull request by modmuss50**
Might be useful if you want to diff the source.

### CODE REVIEW

**Review by liach — APPROVED** [review]
Mind adding this to readme as well.

File: build.gradle [hunk 1]
```diff
@@ -473,6 +476,15 @@ task mapNamedJar(dependsOn: [buildTinyWithEnum, mapIntermediaryJar]) {
...
 }
 
+task decompileCFR(dependsOn: mapNamedJar, type: JavaExec) {
+	doFirst {
+		file('src').deleteDir()
+	}
+
+	classpath = configurations.cfrRuntime
+	main = 'org.benf.cfr.reader.Main'
+	args '--outputdir', 'src', namedJar.getAbsolutePath()
```
Comment (liach) [inline build.gradle:489]: Given we have already get the libraries when we've map the jar, we can just supply the libraries to the decompiler. The flag is `--extraclasspath`, and it seems to take a parameter like java's classpath (a string)

**Review by liach — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
Comment (modmuss50) [reply build.gradle:489]: It seems to decompile just fine without it, ill look into setting that up but I dont think its needed.

Comment (modmuss50) [reply build.gradle:489]: Done.

**Review by liach — APPROVED** [review]

**Comment by modmuss50** [comment]
humm, that last change was bad as it assumes the libs have been downloaded before the task is evaluated, and in the cast of Travis they hadnt

**Comment by liach** [comment]
I think you should have written `{it -> it.name}` instead of just `{it.name}` possibly

**Comment by ChloeDawn** [comment]
`{it.name}` is perfectly valid. Although I would personally add whitespace for readability: `{ it.name }`

**Comment by modmuss50** [comment]
Yeah, the issue is the library’s haven’t been downloaded when the args are processed. I’m going to diff the source with and without the libs to see if it’s even needed.

File: build.gradle [hunk 2]
```diff
@@ -473,6 +479,15 @@ task mapNamedJar(dependsOn: [buildTinyWithEnum, mapIntermediaryJar]) {
...
 }
 
+task decompileCFR(dependsOn: mapNamedJar, type: JavaExec) {
+	doFirst {
+		file('src').deleteDir()
+	}
+
+	classpath = configurations.cfrRuntime
+	main = 'org.benf.cfr.reader.Main'
+	args '--outputdir', 'src', '--extraclasspath', Stream.of(libraries.listFiles()).map{it.name}.collect(Collectors.joining(";")), namedJar.getAbsolutePath()
```
Comment (ChloeDawn) [inline build.gradle]:
```groovy
Stream.of(libraries.listFiles()).map{it.name}.collect(Collectors.joining(";"))
```
could just be
```groovy
libraries.listFiles().collect { it.name }.join ';'
```

**Review by ChloeDawn — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 2 above)
Comment (modmuss50) [reply build.gradle]: yeah, thats better but it doesnt solve the first issue of the libs not being there. ive removed adding the libs as the code looks fine without it (I havent been able to diff it just yet)

**Comment by liach** [comment]
mind updating the base branch the pull request is targetting?

**Review by Runemoro — APPROVED** [review]

**Comment by Runemoro** [comment]
> Undo adding libs as I dont think its required

It is required. Without libs, the decompiler will:

 1. Be forced to add casts to every library method call (`LibraryClass.doSomething((Something) something)`) since it doesn't know that the call isn't ambiguous.
 2. Be unable to correctly infer generics or remove unchecked casts since it needs the function signature which is in the library class.

**Comment by liach** [comment]
now as we always download libs after the intermediary migration, consider adding libs back.

**Pull request closed without merging**